### PR TITLE
added blog SC

### PR DESCRIPTION
**Pull request by windushka**

### CODE REVIEW

File: .github/workflows/ci-tests_blog.yml [hunk 1]
```diff
@@ -0,0 +1,26 @@
+name: Unit tests
```
Comment (qdrn) [inline .github/workflows/ci-tests_blog.yml:1]: I believe the naming of the file is unfortunate. Either you should rename the file that test the helloworld as well. Either you should put everything under `ci-tests.yml`.

File: blog/assembly/__tests__/massa-example.spec.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+// The describe function is used to group related tests together.
+// In this case, we're grouping tests related to the `main` function.
+describe('main function', () => {
+  // The it function is used to define individual test cases.
+  // In this case, we're testing that the `constructor` function doesn't throw an error when called with no arguments.
+  it('should not throw with no arguments', () => {
+    // We're using the expect function to make an assertion about the behavior of the `constructor` function.
+    // Specifically, we're expecting that calling `main` with serialized `constructor` will not throw an error.
+    expect(
```
Comment (qdrn) [inline blog/assembly/__tests__/massa-example.spec.ts:12]: You should test:
- constructor
- post
- deletePost
- _blogKey

File: blog/assembly/contracts/main.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    const post = argsDeser
+    .nextString()
+    .expect('Post argument is missing or invalid');
+    var n_blog_posts = parseInt(Storage.get("N_BLOG_POSTS"));
+    n_blog_posts += 1;
+    Storage.set(_blogKey(n_blog_posts.toString()), post);
+    Storage.set("N_BLOG_POSTS", n_blog_posts.toString());
+}
+
+export function deletePost(binaryArgs: StaticArray<u8>): void {
```
Comment (qdrn) [inline blog/assembly/contracts/main.ts:32]: Should improve documentation

File: blog/assembly/contracts/main.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    const argsDeser = new Args(binaryArgs);
+    const postIndex = argsDeser
+    .nextString()
+    .expect('Post index argument is missing or invalid');
+    Storage.set(_blogKey(postIndex), "");
+}
+
+function _blogKey(postIndex: string): string {
```
Comment (qdrn) [inline blog/assembly/contracts/main.ts:40]: Same

File: blog/assembly/contracts/main.ts [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    if (!callerHasWriteAccess()) {
+        return [];
+      }
+    var n_blog_posts = 0;
+    Storage.set("N_BLOG_POSTS", n_blog_posts.toString());
+    generateEvent(`Blog initiated`);
+    return [];
+}
+
+export function post(binaryArgs: StaticArray<u8>): void {
```
Comment (qdrn) [inline blog/assembly/contracts/main.ts:21]: Should improve documentation

**Review by qdrn — CHANGES_REQUESTED** [review]

**Comment by qdrn** [comment]
Also you should fix the linter!

File: blog/assembly/contracts/main.ts [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+/**
+ * This function is meant to be called only one time: when the contract is deployed.
+ *
+ * @param _ - not used
+ */
+export function constructor(_: StaticArray<u8>): StaticArray<u8> {
+  // This line is important. It ensures that this function can't be called in the future.
+  // If you remove this check, someone could call your constructor function and reset your smart contract.
+  if (!callerHasWriteAccess()) {
```
Comment (gregLibert) [inline blog/assembly/contracts/main.ts:12]: You should be able to use [isDeployingContract](https://github.com/massalabs/massa-as-sdk/blob/main/assembly/std/context.ts#L46) which is cleaner

File: blog/assembly/contracts/main.ts [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+  const nBlogPosts = 0;
+  // Set the initial value of N_BLOG_POSTS to 0 in the storage of the contract
+  Storage.set("N_BLOG_POSTS", nBlogPosts.toString());
+  // This function is used to emit an event to the blockchain
+  generateEvent(`Blog initiated`);
+  return [];
+}
+
+export function post(binaryArgs: StaticArray<u8>): void {
+  const argsDeser = new Args(binaryArgs);
```
Comment (gregLibert) [inline blog/assembly/contracts/main.ts:24]: pratice is to use _args and args:

```suggestion
export function post(_args: StaticArray<u8>): void {
  const args = new Args(_args);
```

File: blog/assembly/contracts/main.ts [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+  // Get the post from the arguments and check if it is valid (not null)
+  const post = argsDeser
```
Comment (gregLibert) [inline blog/assembly/contracts/main.ts:26]:
```suggestion
  const postContent = args
```

File: blog/assembly/contracts/main.ts [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+    .nextString()
+    .expect('Post argument is missing or invalid');
+  let nBlogPosts = parseInt(Storage.get("N_BLOG_POSTS"));
```
Comment (gregLibert) [inline blog/assembly/contracts/main.ts:29]:
```suggestion
  let numberOfPosts = parseInt(Storage.get("N_BLOG_POSTS"));
```

File: blog/package.json [hunk 10]
```diff
@@ -0,0 +1,45 @@
+{
+    "name": "my-massa-sc",
```
Comment (gregLibert) [inline blog/package.json:2]: it should be something around blog and examples no ?

**Review by gregLibert — CHANGES_REQUESTED** [review]
Good job, small improvments to get the code easier to understand and the tool easier to maintain.

File: .github/workflows/ci-tests_blog.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (windushka) [reply .github/workflows/ci-tests_blog.yml]: Done

File: .github/workflows/daily-deploy_blog.yml [hunk 11]
```diff
@@ -0,0 +1,41 @@
+name: Daily Deploy
```
[the remark this answers is not in the document]
Comment (windushka) [reply .github/workflows/daily-deploy_blog.yml]: Done

File: blog/assembly/contracts/main.ts (hunk 6 above)
Comment (gregLibert) [reply blog/assembly/contracts/main.ts:12]: Why was this discussioin resolved ? isDeployingContract is not working ?

File: blog/assembly/contracts/main.ts [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+export function post(_args: StaticArray<u8>): void {
+  const args = new Args(_args);
+  // Get the post from the arguments and check if it is valid (not null)
+  const post = args
+    .nextString()
+    .expect('Post argument is missing or invalid');
+  let numberOfPosts = parseInt(Storage.get("N_BLOG_POSTS"));
+  numberOfPosts += 1;
+  // Store the post in the storage of the contract with the key POST_postIndex
+  // The keys will have the following syntaxes: POST_1.0, POST_2.0, POST_3.0, etc.
```
Comment (gregLibert) [inline blog/assembly/contracts/main.ts:32]: I'm not sure about the .0 in 1.0, 2.0 ... numberOfPosts is an integer so it should be POST_1, POST_2 ...

**Review by gregLibert — CHANGES_REQUESTED** [review]

File: blog/assembly/contracts/main.ts (hunk 12 above)
Comment (windushka) [reply blog/assembly/contracts/main.ts:32]: It is indeed an integer but for some unusual reason numberOfPosts will always have the decimal part even when applying Math.floor() on it. 

image

Comment (windushka) [reply blog/assembly/contracts/main.ts:32]: For isDeployingContract() to work I had to update to the latest massa-as-sdk version which resulted in my unit tests not working. When readingthe storage the returned value would always be NaN. I've addressed this issue to the run team.

Comment (gregLibert) [reply blog/assembly/contracts/main.ts:32]: This is not a priority. You can safely ignore my comment

Comment (windushka) [reply blog/assembly/contracts/main.ts]: Using [isDeployingContract](https://github.com/massalabs/massa-as-sdk/blob/main/assembly/std/context.ts#L46) now and fixed the decimal part with the help of @Ben-Rey

File: blog/assembly/contracts/main.ts [hunk 13]
```diff
@@ -0,0 +1,55 @@
+import { Storage, Context, generateEvent } from '@massalabs/massa-as-sdk';
+import { Args } from '@massalabs/as-types';
+
+/**
+ * This function is meant to be called only one time: when the contract is deployed.
+ *
+ * @param _ - not used
+ */
+export function constructor(_: StaticArray<u8>): StaticArray<u8> {
```
Comment (gregLibert) [inline blog/assembly/contracts/main.ts:9]:
```suggestion
export function constructor(_: StaticArray<u8>): void {
```

File: blog/assembly/contracts/main.ts [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+  // This line is important. It ensures that this function can't be called in the future.
+  // If you remove this check, someone could call your constructor function and reset your smart contract.
+  if (!Context.isDeployingContract()) {
+    return [];
```
Comment (gregLibert) [inline blog/assembly/contracts/main.ts:13]:
```suggestion
    return;
```

File: blog/assembly/contracts/main.ts [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+  }
+  const nBlogPosts = 0;
+  // Set the initial value of N_BLOG_POSTS to 0 in the storage of the contract
+  Storage.set("N_BLOG_POSTS", nBlogPosts.toString());
+  // This function is used to emit an event to the blockchain
+  generateEvent('Blog initiated');
+  return [];
```
Comment (gregLibert) [inline blog/assembly/contracts/main.ts:20]:
```suggestion
```

File: blog/assembly/contracts/main.ts [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+export function post(_args: StaticArray<u8>): void {
+  const args = new Args(_args);
+  // Get the post from the arguments and check if it is valid (not null)
+  const post = args
+    .nextString()
+    .expect('Post argument is missing or invalid');
+
+  let numberOfPosts = 0;
```
Comment (gregLibert) [inline blog/assembly/contracts/main.ts:30]:
```suggestion
  let postLastIndex = 0;
```

File: blog/assembly/contracts/main.ts [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+  if (Storage.get<string>("N_BLOG_POSTS") !== "") {
+    numberOfPosts = parseInt(Storage.get("N_BLOG_POSTS")) as i32;
```
Comment (gregLibert) [inline blog/assembly/contracts/main.ts:32]:
```suggestion
    postLastIndex = parseInt(Storage.get("N_BLOG_POSTS")) as i32;
```

File: blog/assembly/contracts/main.ts [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+  }
+  numberOfPosts += 1;
```
Comment (gregLibert) [inline blog/assembly/contracts/main.ts:34]:
```suggestion
  postLastIndex += 1;
```

File: blog/assembly/contracts/main.ts [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+  // Store the post in the storage of the contract with the key POST_postIndex
+  // The keys will have the following syntaxes: POST_1, POST_2, POST_3, etc.
+  Storage.set(blogKey(numberOfPosts.toString()), post);
```
Comment (gregLibert) [inline blog/assembly/contracts/main.ts:37]:
```suggestion
  Storage.set(blogKey(postLastIndex.toString()), post);
```

File: blog/assembly/contracts/main.ts [hunk 20]
```diff
@@ -0,0 +1,55 @@
...
+  // Incrementing the value of N_BLOG_POSTS in the storage of the contract
+  Storage.set("N_BLOG_POSTS", numberOfPosts.toString());
```
Comment (gregLibert) [inline blog/assembly/contracts/main.ts:39]:
```suggestion
  Storage.set("N_BLOG_POSTS", postLastIndex.toString());
```

**Review by gregLibert — CHANGES_REQUESTED** [review]

**Comment by windushka** [comment]
Addressed all the requested changes and added one max parallel job in the matrix strategy. When there were multiple deployment jobs running in parallel, sometimes the event wouldn't be received. Should be fixed now

**Review by gregLibert — APPROVED** [review]

**Pull request merged**